lib/flight_scheduler/schedulers/backfilling_scheduler.rb

```diff
@@ -0,0 +1,261 @@
+      # over the jobs.
+      candidates
+        .take(5)
+        .select { |job| job.reason_pending == 'WaitingForScheduling' }
+        .each { |job| job.reason_pending = 'Priority' }
+    end
+    new_allocations
+  end
+
+  def queue
+    # For the backfilling scheduler, the queue is sorted and slightly filtered
+    # view of the registry of jobs.
+    #
+    # The sorting and filtering is primarily intended for display purposes,
+    # but also using it for processing ensures that the jobs are processed in
+    # the order in which the queue output suggests.
+    #
+    # The jobs are sorted to ensure that running jobs are placed higher in the
+    # queue output than pending jobs.
+    #
+    # The jobs are filtered to ensure that array jobs without any pending
+    # tasks are not included.
+    #
+    # A more complicated scheduler may implement a priority queue and need to
+    # keep its own record of the queued jobs rather than using job registry.
+    running_jobs_first = lambda { |job| job.reason_pending.nil? ? -1 : 1 }
+
+    # NOTE: We rely on the sort being stable to ensure that earlier added jobs
+    # are considered prior to later added jobs.
+    FlightScheduler.app.job_registry.jobs
+      .reject { |j| j.job_type == 'ARRAY_JOB' && j.task_generator.finished? }
+      .reject { |j| j.terminal_state? }
+      .sort_by.with_index { |x, idx| [running_jobs_first.call(x), idx] }
+  end
+
+  # Attempt to allocate a single job.
+  #
+  # If the partition has sufficient resources available for the job, a
+  # new +Allocation+ is added to the allocations registry and returned.
+  #
+  # If an allocation is created for an `ARRAY_TASK`, it is added to the job
+  # registry and the associated task generator advanced.
+  #
+  # If there are insufficient resources available to allocate to the job,
+  # the job's `reason_pending` is updated and +nil+ is returned.
+  #
+  # NOTE: It is expected that this will not differ between schedulers.
+  def allocate_job(job, reason: 'Resources')
+    raise InvalidJobType, job if job.job_type == 'ARRAY_JOB'
+
+    # Generate an allocation for the job
+    nodes = job.partition.nodes
+    FlightScheduler::LoadBalancer.new(job: job, nodes: nodes).allocate.tap do |allocation|
+      if allocation.nil?
+        if job.job_type == 'ARRAY_TASK'
+          job.array_job.reason_pending = reason
+        else
+          job.reason_pending = reason
+        end
+        nil
+      else
+        if job.job_type == 'ARRAY_TASK'
+          FlightScheduler.app.job_registry.add(job)
+          job.array_job.task_generator.advance_next_task
+        end
+        FlightScheduler.app.allocations.add(allocation)
+      end
+    end
+  end
+
+  def candidates
+    # The maximum number of queued jobs to consider.
+    max_jobs_to_consider = 50
+    considered = 0
+
+    Enumerator.new do |yielder|
+      queue.each do |job|
+        considered += 1
+        if considered > max_jobs_to_consider
```

Comment:
I'm not sure about `max_jobs_to_consider`, I think it may cause issues in the future. It is basically saying only look at the first 50 jobs of the queue, even if they are all invalid for some reason.

I wonder what would happen if the first 50 jobs all exceed the partition time limit but the 51st job doesn't? The enumerator would never be able to reach it, and thus lock the scheduler.

So I have two queries about this:

1. Is there a particular need for a `max_jobs_to_consider`? The scheduler will only call the `Enumerator` as many times it needs to. So there is very little harm letting it run its cause
2. If their needs to be a maximum, then the count should be post hoc. It should only count the "valid" candidates it yields and skip over the un-runnable jobs.

Reply:
(2) seems like a much better idea than what I'm doing here.  Thanks.